kmk/extensions/oled.py

```diff
@@ -0,0 +1,171 @@
+
+DISPLAY_OFFSET = 4  # Used to calculate a new zero level since SSD1306 is clipped.
+
+
+class OledEntryType:
+    TXT = 0
+    IMG = 1
+
+
+class OledData:
+    def __init__(
+        self,
+        entries=None,
+    ):
+        if entries is not None:
+            self.data = entries
+
+    @staticmethod
+    def oled_text_entry(x=0, y=0, text='', layer=None):
+        return {
+            0: text,
+            1: x,
+            2: y,
+            3: layer,
+            4: OledEntryType.TXT,
+        }
+
+    @staticmethod
+    def oled_image_entry(x=0, y=0, image='', layer=None):
+        odb = displayio.OnDiskBitmap(image)
+        return {
+            0: odb,
+            1: x,
+            2: y,
+            3: layer,
+            4: OledEntryType.IMG,
+        }
+
+
+class Oled(Extension):
+    def __init__(
+        self,
+        views,
+        width=128,
+        height=32,
+        flip: bool = False,
+        device_address=0x3C,
+        brightness=0.8,
+        brightness_step=0.1,
+    ):
+        displayio.release_displays()
+        self.rotation = 180 if flip else 0
+        self._views = views.data
+        self._width = width
+        self._height = height
+        self._prevLayers = 0
+        self._device_address = device_address
+        self._brightness = brightness
+        self._brightness_step = brightness_step
+        gc.collect()
+
+        make_key(
+            names=('OLED_BRI',), on_press=self._oled_bri, on_release=handler_passthrough
+        )
+        make_key(
+            names=('OLED_BRD',), on_press=self._oled_brd, on_release=handler_passthrough
+        )
+
+    def render_oled(self, layer):
+        splash = displayio.Group()
+
+        for view in self._views:
+            if view[3] == layer or view[3] is None:
+                if view[4] == OledEntryType.TXT:
+                    splash.append(
+                        label.Label(
+                            terminalio.FONT,
+                            text=view[0],
+                            color=0xFFFFFF,
+                            x=view[1],
+                            y=view[2] + DISPLAY_OFFSET,
+                        )
+                    )
+                elif view[4] == OledEntryType.IMG:
+                    splash.append(
+                        displayio.TileGrid(
+                            view[0],
+                            pixel_shader=view[0].pixel_shader,
+                            x=view[1],
+                            y=view[2] + DISPLAY_OFFSET,
+                        )
+                    )
+        gc.collect()
+        self._display.show(splash)
+
+    def updateOLED(self, sandbox):
+        self.render_oled(sandbox.active_layers[0])
+        gc.collect()
+
+    def on_runtime_enable(self, sandbox):
+        return
+
+    def on_runtime_disable(self, sandbox):
+        return
+
+    def during_bootup(self, board):
+        displayio.release_displays()
+        i2c = busio.I2C(board.SCL, board.SDA)
```

Comment:
```suggestion
        i2c = busio.I2C(keyboard.SCL, keyboard.SDA)
```
We define all the pins in keyboard class anyway, this way we can use I2C pins that aren't default.